pallas-crypto/src/nonce/mod.rs

```diff
@@ -0,0 +1,17 @@
+use thiserror::Error;
+
+use crate::hash::Hash;
+
+pub mod epoch_nonce;
+pub mod rolling_nonce;
+
+#[derive(Error, Debug)]
+pub enum Error {
+    #[error("Nonce error: {0}")]
+    Nonce(String),
+}
+
+/// A trait for generating nonces.
+pub trait NonceGenerator: Sized {
+    fn finalize(&mut self) -> Result<Hash<32>, Error>;
+}
```

Comment:
Can we change a bit the trait and remove this Error type that is not carrying any useful meaning in this case ?

How about going to a more generic approach:

```rust
pub trait NonceGenerator {
    type Output;

     fn finalize(&mut self) -> Self::Output;
 }
```

That way we can remove the error handling that is not necessary for the EpochGenerator:

```rust
impl NonceGenerator for EpochNonceGenerator {
    type Output = Hash<32>;

     fn finalize(&mut self) -> Self::Output {
         Ok(self.nonce)
     }
 }
```

And have an error handling more precise for the `RollingNonceGenerator`:

```rust
#[derive(Error)]
#[error("Must call apply_block before finalise!")]
struct RollingNonceGeneratorError;

impl NonceGenerator for RollingNonceGenerator {
    type Output = Result<Hash<32>, RollingNonceGeneratorError>;

     fn finalize(&mut self) -> Self::Output {
        todo!()
     }
 }
```

What do you think? Would that make sense on the overall picture ? I like it in the sense that (though I recognise I am bias for this kind of API) it gives more explicit details of what the Error handling is (and no error handling if not needed) and also allows the opportunity to remove the need to stick to `Hash<32>` (allow to use `Hash<24>` for example).

Reply:
Do we need a trait at all?

Reply:
Indeed, I don't think this is needed.

Reply:
I'll remove and keep it simple with just some functions.

Reply:
This is ready to re-check.